### PR TITLE
Advanced version for proxy requests

### LINKED ISSUE #86: [proxy] proxy requests instead of redirecting
**Is this a BUG REPORT or FEATURE REQUEST?**:
feature

**What happened**:
Currently we redirect the user, but don't enable proxying requests.

**What you expected to happen**:
Ability to proxy to url instead of redirect.

Something to allow `demo.example.com` to be served from `example.com/demo` or `google.com` at `example.com`

**Anything else we need to know?**:
* whitelist/blacklist urls (to prevent abuse).
* could be a new type or a specific config for host/path

Needs more discussion, but would be interesting.

### PR DESCRIPTION
**Pull request by dbriesz**
**What this PR does / why we need it**:
Allows for more advanced proxy request configuration such as configuring multiple upstreams and adding headers to the request.

**Which issue this PR fixes**:
Extends #86

### CODE REVIEW

File: caddy/caddy.go [hunk 1]
```diff
@@ -77,6 +79,16 @@ func parse(c *caddy.Controller) (txtdirect.Config, error) {
 			}
 			resolver = resolverAddr[0]
 
+		case "proxy":
+			toProxy := c.RemainingArgs()
```
Comment (stp-ip) [inline caddy/caddy.go]: Still not sure, what proxy config you thought should work here.

For something like:
```
txtdirect {
  enable proxy
  resolver 127.0.0.1
  proxy / example.com
}
```
The `toProxy` would result in ["/", "example.com"]

File: caddy/caddy.go [hunk 2]
```diff
@@ -77,6 +79,16 @@ func parse(c *caddy.Controller) (txtdirect.Config, error) {
...
+			if len(toProxy) != 1 {
```
Comment (stp-ip) [inline caddy/caddy.go]: Not sure, why you want to check for this specific length. Could you give a bit more context on what you wanted to do?

File: caddy/caddy.go [hunk 3]
```diff
@@ -77,6 +79,16 @@ func parse(c *caddy.Controller) (txtdirect.Config, error) {
...
+				return txtdirect.Config{}, c.ArgErr()
+			}
+			upstreams, err := proxy.NewStaticUpstreams(c.Dispenser, toProxy[0])
```
Comment (stp-ip) [inline caddy/caddy.go]: c.Dispenser will error out here as it doesn't get an actual proxy configuration. Let's remove the config logic first and use something hardcoded:
```
upstreams, err := proxy.NewStaticUpstreams(caddyfile.NewDispenser("Testfile", strings.NewReader(`
			proxy / example.com
		`)), "")
```

What that basically does:
* don't set a host (simplification)
* create a new dispenser with the most simple configuration

File: caddy/caddy.go (hunk 2 above)
Comment (dbriesz) [reply caddy/caddy.go]: I was thinking I could check if the proxy config contains 1 token as a starting point with a simple example. I changed this to return an error if the proxy arguments in the config file are empty instead.

File: caddy/caddy.go (hunk 3 above)
Comment (dbriesz) [reply caddy/caddy.go]: I added a hardcoded example to start. Here is my local config that I used to test it:

In the txt record:
```
_redirect.two.one     60 IN TXT     "v=txtv0;to=http://google.com;type=proxy;code=302"
```

In the caddyfile:
```
example.test:8080 {
  tls off
  txtdirect {
    enable path host proxy
    resolver 127.0.0.1:5353
    proxy / google.com
    logfile stdout
  }
  log / stdout "{remote} - [{when_iso}] \"{method} {uri} {proto}\" {status} {size} {latency}"
  errors stdout
}
```

**Comment by stp-ip** [comment]
For now I would close this and reiterate, when we have some better ideas on how this should work.

**Pull request closed without merging**